src/live/src/user-to-user/stream.js

```diff
@@ -0,0 +1,245 @@
+   */
+  stopListening() {
+    const userId = this.client.activeUser._id;
+    return this._unsubscribe(userId);
+  }
+
+  /**
+   * Sends a message to the specified user
+   * @param {string} userId
+   * @param {Object} message
+   */
+  send(userId, message) {
+    return this._publish(userId, message);
+  }
+
+  /**
+   * @private
+   * @param {string} substreamOwnerId
+   * @returns {Promise}
+   */
+  _unsubscribeFromSubstream(substreamOwnerId) {
+    const path = `${substreamOwnerId}/unsubscribe`;
+
+    return this._makeStreamRequest(path, RequestMethod.POST, { deviceId: this.client.deviceId })
+      .then((response) => {
+        delete this.subscribeChannels[substreamOwnerId];
+        return response;
+      });
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @param {MessageReceiver} receiver
+   * @returns {Promise}
+   */
+  _subscribe(userId, receiver) {
+    return this._requestSubscribeAccess(userId)
+      .then(() => {
+        const channelName = this.subscribeChannels[userId] || this._buildChannelName(userId);
+        return this.liveService.subscribeToChannel(channelName, receiver);
+      });
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @returns {Promise<void>}
+   */
+  _unsubscribe(userId) {
+    return this._unsubscribeFromSubstream(userId)
+      .then(() => {
+        const channelName = this.subscribeChannels[userId] || this._buildChannelName(userId);
+        this.liveService.unsubscribeFromChannel(channelName);
+      });
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @param {Object} message
+   * @returns {Promise}
+   */
+  _publish(userId, message) {
+    if (this.publishChannels[userId]) {
+      return this._retriablePublish(userId, message);
+    }
+    return this._publishWithAccessRequest(userId, message);
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @param {Object} message
+   * @returns {Promise}
+   */
+  _publishWithAccessRequest(userId, message) {
+    return this._requestPublishAccess(userId)
+      .then(() => {
+        const channelName = this.publishChannels[userId] || this._buildChannelName(userId);
+        return this.liveService.publishToChannel(channelName, message);
+      });
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @param {Object} message
+   * @returns {Promise}
+   */
+  _retriablePublish(userId, message) {
+    const channelName = this.publishChannels[userId] || this._buildChannelName(userId);
+    return this.liveService.publishToChannel(channelName, message)
+      .catch((err) => {
+        let promise = Promise.reject(err);
+        if (err.statusCode === StatusCode.Forbidden) {
+          promise = this._publishWithAccessRequest(userId, message);
+        }
+        return promise;
+      });
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @returns {string} The name of the channel for that user
+   */
+  _buildChannelName(userId) {
```

Comment:
Now it's only used as a fallback.

Initially, I wanted to directly publish and if I got a 403, only then call `/publish`. This way, you'd save one HTTP call in cases where you've still got permission from your last call to `/publish`, but cause of the caching 403 issue I talked about, this didn't work well, because it would cache that 403 for 30 seconds, despite you requesting and being granted access. If that had remained the flow, I'd need to build the channel name the first time. If it wasn't for that 403 issue, I would still prefer this method, but now that we know this can happen, the cautious way seems better, at least until we know why/when it happens.